### PR TITLE
LPS-143885 Move remaining files and tag from frontend-taglib-clay to a dedicate module

### PR DESCRIPTION
**Pull request by LuismiBarcos**
Related ticket -> [LPS-143885](https://issues.liferay.com/browse/LPS-143885)

The purpose of this branch is to move the logic of the data-set-display from the module `frontend-taglib-clay` to `frontend-data-set` 

I copied, pasted and renamed the remaining files from frontend-taglib-clay to frontend-data-set. I set the files in the module I consider that it belongs (-api, -impl, -taglib...).

To test this I migrate two modules (not in this PR) and, doing that, I realized that other teams are adding more logic to the module `frontend-taglib-clay` doing the migration more difficult.

As we are going to do the migration of the components that are using these tags, we can take care of migrating these new files but, at the time we start to do the migration, I suppose that we have to tell the other teams that the module `frontend-data-set`  will be the new place for the logic of the data set display.

### CODE REVIEW

**Comment by javierdearcos** [comment]
> To test this I migrate two modules (not in this PR) and, doing that, I realized that other teams are adding more logic to the module frontend-taglib-clay doing the migration more difficult.

You're completely right. We need to talk with the other teams to freeze the improvements and new features until the migration is done.

Thank you for bring this to light

**Review by javierdearcos — APPROVED** [review]
LGTM :)

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/frontend-data-set.tld [hunk 1]
```diff
@@ -10,6 +10,130 @@
 	<tlib-version>1.0</tlib-version>
 	<short-name>frontend-data-set</short-name>
 	<uri>http://liferay.com/tld/frontend-data-set</uri>
+	<tag>
+		<name>data-set-display</name>
+		<tag-class>com.liferay.frontend.data.set.taglib.servlet.taglib.DataSetDisplayTag</tag-class>
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/frontend-data-set.tld:15]: Oh, you're migrating the remaining tag as well! :open_mouth:  Awesome!

My plan was to name the tag `<frontend-data-set:classic-display>`. I think it's a bit awkward to have double `data-set`: `<frontend-data-set:data-set-display>`, and we want to distance ourselves from `data-set-display` phrase. That should be left for the old component, to prevent confusion in discussions. Along with the change in this file, this would mean:
- tag class should be `ClassicDisplayTag.java`
- jsp dir should be `classic_display`
- attributes should be `frontend-data-set:classic-display:*`

Of course, `classic-display` is not set in stone, but it should be a name that rhymes with `headless-display`.

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/DataSetDisplayTag.java [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.jsp.JspException;
+import javax.servlet.jsp.PageContext;
+
+/**
+ * @author Marco Leo
+ * @author Alessio Antonio Rendina
+ */
+public class DataSetDisplayTag extends IncludeTag {
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/DataSetDisplayTag.java:55]: [Not a blocker for PR] Since API of two tags are very similar, my plan here was to make a `BaseDisplayTag` class that this one and `HeadlessDisplayTag` would extend. But, we can leave that for a separate PR.

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/internal/jaxrs/application/FDSApplication.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+import com.liferay.frontend.data.set.provider.FDSDataProvider;
+import com.liferay.frontend.data.set.provider.FDSDataProviderRegistry;
+import com.liferay.frontend.data.set.provider.search.FDSPagination;
+import com.liferay.frontend.data.set.taglib.internal.jaxrs.context.provider.PaginationContextProvider;
+import com.liferay.frontend.data.set.taglib.internal.jaxrs.context.provider.SortContextProvider;
+import com.liferay.frontend.data.set.taglib.internal.jaxrs.context.provider.ThemeDisplayContextProvider;
+import com.liferay.frontend.data.set.taglib.internal.servlet.ServletContextUtil;
+import com.liferay.frontend.taglib.clay.data.Filter;
+import com.liferay.frontend.taglib.clay.data.FilterFactory;
+import com.liferay.frontend.taglib.clay.data.FilterFactoryRegistry;
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/internal/jaxrs/application/FDSApplication.java:27]: These should be:
```java
com.liferay.frontend.data.set.provider.search.FDSKeywords;
com.liferay.frontend.data.set.provider.search.FDSKeywordsFactory;
com.liferay.frontend.data.set.provider.search.FDSKeywordsFactoryRegistry;
```
If you are wondering why we changed this so much: the naming of classes in `frontend.taglib.clay` made no sense for their purpose. [FDSFilter](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/filter/FDSFilter.java) is a model classes for UI component, looking misleadingly similar.

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/DataSetDisplayTag.java [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+
+package com.liferay.frontend.data.set.taglib.servlet.taglib;
+
+import com.liferay.frontend.data.set.model.FDSPaginationEntry;
+import com.liferay.frontend.data.set.taglib.internal.servlet.ServletContextUtil;
+import com.liferay.frontend.data.set.view.FDSViewSerializer;
+import com.liferay.frontend.taglib.clay.servlet.taglib.util.CreationMenu;
+import com.liferay.frontend.taglib.clay.servlet.taglib.util.DropdownItem;
+import com.liferay.frontend.taglib.clay.servlet.taglib.util.SortItem;
+import com.liferay.frontend.taglib.clay.servlet.taglib.util.SortItemList;
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/DataSetDisplayTag.java:23]: These should be:
```java
import com.liferay.frontend.data.set.model.FDSSortItem;
import com.liferay.frontend.data.set.model.FDSSortItemList;
```

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/resources/data_set/init.jsp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+ */
+--%>
+
+<%@ taglib uri="http://liferay.com/tld/react" prefix="react" %><%@
+taglib uri="http://liferay.com/tld/theme" prefix="liferay-theme" %>
+
+<%@ page import="com.liferay.frontend.data.set.model.FDSPaginationEntry" %><%@
+page import="com.liferay.frontend.taglib.clay.servlet.taglib.util.CreationMenu" %><%@
+page import="com.liferay.frontend.taglib.clay.servlet.taglib.util.DropdownItem" %><%@
+page import="com.liferay.frontend.taglib.clay.servlet.taglib.util.SortItemList" %><%@
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/resources/data_set/init.jsp:23]: This should be `FDSSortItemList`

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/frontend-data-set.tld [hunk 6]
```diff
@@ -10,6 +10,130 @@
...
+			<required>false</required>
+			<rtexprvalue>true</rtexprvalue>
+		</attribute>
+		<attribute>
+			<name>showSearch</name>
+			<required>false</required>
+			<rtexprvalue>true</rtexprvalue>
+		</attribute>
+		<attribute>
+			<name>sortItemList</name>
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/frontend-data-set.tld:127]: This attribute should be `fdsSortItemList`, like in `headless-display` tag.

**Review by markocikos — CHANGES_REQUESTED** [review]
@LuismiBarcos Awesome work! I'm very glad you're on this :hugs:  I do think some tweaks are still needed, please see comments inline.

**Comment by markocikos** [comment]
@LuismiBarcos This PR will have a minor semver conflict with https://github.com/liferay-objects/liferay-portal/pull/298. This is fine, I don't think we should stop progress on either. Just have in mind, after one is merged, the other will need to update semver.

**Comment by LuismiBarcos** [comment]
I've applied the changes requested by @markocikos (see previous discussions for more info) and update the commits.
As a resume:
- I've deleted the classes `FDSContentRendererContextContributor`,  `FDSContentRendererContextContributorRegistry` and `FDSContentRendererContextContributorRegistryImpl` because, as Marko said, there was repeated logic.
- I've moved the class `FDSDataRow` to the `-api` module to be accessible by the rest of the modules. With that, I was able to delete the `FDSDataJsonFactory` from the `-api` and its implementation from the `-impl` module.
- The logic of `FDSDataJsonFactory` was added to the  `-taglib` module inside the internal package `factory`. To make it works I had to add more classes that implement the FDSKeywords* interfaces to provide services in OSGi that resolves de dependencies (something expected)
- I've moved `FDSActionProviderRegistryImpl` to the package `provider` 
- I've renamed the tag to `classic-display`
- I've updated all the necessary imports to use the FDS* classes

I hope I explained myself 😅 
Also, after all these changes y tested again doing migrations for object-web and commerce-catalog, and works for me. If you can review again to ensure that I've not made any error I will really appreciate it 🙂

File: modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/FDSContentRendererContextContributor.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * @author Marco Leo
+ */
+public interface FDSContentRendererContextContributor {
+
+	public Map<String, Object> getContentRendererContext(
+		FDSView fdsView, Locale locale);
```
[the remark this answers is not in the document]
Comment (LuismiBarcos) [reply modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/FDSContentRendererContextContributor.java]: Done, I've deleted all these classes because the logic was migrated as you said 🙂

File: modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/FDSContentRendererContextContributorRegistry.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+import java.util.List;
+
+/**
+ * @author Marco Leo
+ */
+public interface FDSContentRendererContextContributorRegistry {
+
+	public List<FDSContentRendererContextContributor>
+		getFDSContentRendererContextContributors(String fdsContentRendererName);
```
[the remark this answers is not in the document]
Comment (LuismiBarcos) [reply modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/FDSContentRendererContextContributorRegistry.java]: Done!

File: modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/FDSDataJSONFactory.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+package com.liferay.frontend.data.set;
+
+import java.util.List;
+
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * @author Marco Leo
+ */
+public interface FDSDataJSONFactory {
```
[the remark this answers is not in the document]
Comment (LuismiBarcos) [reply modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/FDSDataJSONFactory.java]: I've moved the logic of FDSDataJsonFactory to the module `-taglib`. To do that, I had to move FDSDataRow to `-api` to make it accessible by the rest of the modules (we do not want a dependency on the `-impl` module).

About `FrontendClayApplication`, this declares some endpoints with `Jaxrs`. The tag `data-set-display` uses custom endpoints declared in this class. As we are migrating the logic without any change, we are algo migrating these endpoints to localize all the logic in `frontend-data-set`. Maybe in the future, if combine these two tags in one, we can avoid these endpoints and use only the headless apis, but, for now, I think it is better to migrate it 🙂 

In fact, if we do not use this endpoints, the tag will always show an empty result, so, it is necessary.

File: modules/apps/frontend-data-set/frontend-data-set-impl/src/main/java/com/liferay/frontend/data/set/internal/FDSActionProviderRegistryImpl.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.data.set.internal;
```
[the remark this answers is not in the document]
Comment (LuismiBarcos) [reply modules/apps/frontend-data-set/frontend-data-set-impl/src/main/java/com/liferay/frontend/data/set/internal/FDSActionProviderRegistryImpl.java]: Done! Thanks 🙂

File: modules/apps/frontend-data-set/frontend-data-set-impl/src/main/java/com/liferay/frontend/data/set/internal/FDSContentRendererContextContributorRegistryImpl.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+import org.osgi.service.component.annotations.Deactivate;
+
+/**
+ * @author Marco Leo
+ */
+@Component(
+	immediate = true,
+	service = FDSContentRendererContextContributorRegistry.class
+)
+public class FDSContentRendererContextContributorRegistryImpl
```
[the remark this answers is not in the document]
Comment (LuismiBarcos) [reply modules/apps/frontend-data-set/frontend-data-set-impl/src/main/java/com/liferay/frontend/data/set/internal/FDSContentRendererContextContributorRegistryImpl.java]: Done!

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/frontend-data-set.tld (hunk 1 above)
Comment (LuismiBarcos) [reply modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/frontend-data-set.tld]: Honestly, I did not have any suggestions for the name. I gave a new name to the tag and waited if the name made sense to you and Brian 😅 

So, I applied the new names that you suggest to this tag in the new commits 🙂

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/DataSetDisplayTag.java (hunk 2 above)
Comment (LuismiBarcos) [reply modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/DataSetDisplayTag.java]: I didn't do this because right now we are not focused on improving the logic (except for special cases, of course) but only in migrate it. I think that it is better to leave it for a separate PR when we finish the migration 🙂

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/internal/jaxrs/application/FDSApplication.java (hunk 3 above)
Comment (LuismiBarcos) [reply modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/internal/jaxrs/application/FDSApplication.java]: Done! thanks for the explanation 🙂

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/DataSetDisplayTag.java (hunk 4 above)
Comment (LuismiBarcos) [reply modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/java/com/liferay/frontend/data/set/taglib/servlet/taglib/DataSetDisplayTag.java]: Done

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/resources/data_set/init.jsp (hunk 5 above)
Comment (LuismiBarcos) [reply modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/resources/data_set/init.jsp]: Done! good catch

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/frontend-data-set.tld (hunk 6 above)
Comment (LuismiBarcos) [reply modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/frontend-data-set.tld]: Done! good catch 🙂

**Comment by javierdearcos** [comment]
Great work @LuismiBarcos! 

And thank you very much @markocikos for the deep review 🙂

File: modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/frontend-data-set.tld (hunk 1 above)
Comment (carloslancha) [reply modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/frontend-data-set.tld]: Hey guys, just leaving my two cents here. I think we shouldn't mix moving all this with renaming stuff since it would make it more difficult to understand the migration when happens. AFAIK the goal of this first phase is just to move it, not improving nor refactor. Once is done we can talk about what we want to have in the end (which will probably be just one tag).

One step after another :)

Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-taglib/src/main/resources/META-INF/frontend-data-set.tld]: @carloslancha I agree with you that we should avoid making any improvements unrelated to migration, as a part of the migration. See for example https://github.com/liferay-frontend/liferay-portal/pull/1766#discussion_r771527102

Having said that, rebuilding the component into a set of new modules is the key part of the migration. When we started on this, we could have made a choice to split migration into (1) copy and rename and (2) rebuild into set of modules. It was my call to skip the first step, for better or worse. It would maybe made migration easier for us on technical level, but longer. I figured we can handle this, and I wanted the code in 'final' form to iterate with Brian to the real final form. 

The toughest part in migration will be updates in product teams, and that will be the same unrelated to either approach.

**Review by markocikos — APPROVED** [review]
LGTM! @LuismiBarcos please update semver, feel free to forward after that.

**Comment by LuismiBarcos** [comment]
ci:forward

**Comment by markocikos** [comment]
@LuismiBarcos CI will fail on semver. Please run `gradle baseline` on `frontend-data-set-api`.

**Comment by LuismiBarcos** [comment]
Oops! I should read better next time 😅

**Comment by LuismiBarcos** [comment]
ci:forward
